### PR TITLE
Eng 19262

### PR DESCRIPTION
**Pull request by manjujames**
- We no longer open a file for writing right at the start. We only create the file when the first write comes in.
- PBD filenames only have that segment's id now. No longer linked to previous segment through filename.
- Start id and end id values may be any long value other than Long.MIN_VALUE. The min_value
indicates that it was never set on the PBD.
- requiresId attribute for PBD is now specified at PBD creation time, instead of it being inferred
on the first offer.

### CODE REVIEW

File: src/frontend/org/voltdb/utils/PersistentBinaryDeque.java [hunk 1]
```diff
@@ -841,37 +808,16 @@ private void parseFiles(boolean deleteExisting) throws IOException {
...
-                        // This is an Id in the sequence referring to a previous file that
-                        // was deleted, so move on.
-                        continue;
-                    }
-                    recoverSegment(index++, segmentId, segmentName);
+            long nextIndex = 1;
+            Long prevIndex = null;
+            for (Map.Entry<Long, PbdSegmentName> entry : filesById.entrySet()) {
+                long currId = entry.getKey();
+                long currIndex = (m_requiresId) ? nextIndex++ : currId;
```
Comment (rdykiel) [inline src/frontend/org/voltdb/utils/PersistentBinaryDeque.java]: Why is this necessary? Can't we use the id extracted from the filename as the index? We just have to check that the filename index matches the segment id.

File: src/frontend/org/voltdb/utils/PersistentBinaryDeque.java [hunk 2]
```diff
@@ -979,7 +925,7 @@ private void quarantineSegment(Map.Entry<Long, PBDSegment<M>> prevEntry, PBDSegm
     }
 
     private PBDSegment<M> findSegmentWithEntry(long entryId, SeekErrorRule errorRule) throws NoSuchOffsetException, IOException {
-        long invalidKey = m_segments.firstKey()-1;
+        long invalidKey = m_segments.isEmpty() ? 1 : m_segments.firstKey()-1;
```
Comment (rdykiel) [inline src/frontend/org/voltdb/utils/PersistentBinaryDeque.java]: Isn't it dangerous to use "1" as an invalid key? Should we use INVALID_ID instead?

File: src/frontend/org/voltdb/utils/PersistentBinaryDeque.java [hunk 3]
```diff
@@ -1607,9 +1516,7 @@ public void close() throws IOException {
 
     @Override
     public synchronized long getFirstId() throws IOException {
-        PBDSegment<M> first = m_segments.firstEntry().getValue();
-        assert(first != null);
-        return first.getStartId();
+        return (m_segments.size() == 0) ? -1 : m_segments.firstEntry().getValue().getStartId();
```
Comment (rdykiel) [inline src/frontend/org/voltdb/utils/PersistentBinaryDeque.java]: Use INVALID_ID?

File: src/frontend/org/voltdb/utils/PBDRegularSegment.java [hunk 4]
```diff
@@ -679,9 +679,9 @@ boolean hasAllFinishedReading() throws IOException {
 
     // Used by Export path
     @Override
-    int offer(DBBPool.BBContainer cont, long startId, long endId) throws IOException
+    int offer(DBBPool.BBContainer cont, Long startId, Long endId) throws IOException
```
[the remark this answers is not in the document]
Comment (manjujames) [reply src/frontend/org/voltdb/utils/PBDRegularSegment.java]: Only E3 uses ids. The rest use the calls without ids.

File: src/frontend/org/voltdb/utils/PbdSegmentName.java [hunk 5]
```diff
@@ -34,12 +34,12 @@
  * </ul>
  */
 public final class PbdSegmentName {
-    // 1(nonce) + 1(_) + 10(currCounter) + 1(_) + 10(prevCounter) + 4(.pbd)
-    private static final int MINIMUM_LENGTH = 27;
+    // 1(nonce) + 1(_) + 10(currCounter) + 1(_) + 4(.pbd)
```
[the remark this answers is not in the document]
Comment (manjujames) [reply src/frontend/org/voltdb/utils/PbdSegmentName.java]: Actually, I think we have to increase it for PBDs where this will be the id because we are storing a long value here. Long values can be > 10^10. However, this is unnecessary for PBDs that don't store ids. Let me think about this and see if I can use 2 different mechanisms depending on which scheme it is.

File: src/frontend/org/voltdb/utils/PersistentBinaryDeque.java [hunk 6]
```diff
@@ -197,6 +200,9 @@ public void seekToSegment(long entryId, SeekErrorRule errorRule)
             synchronized(PersistentBinaryDeque.this) {
                 PBDSegment<M> seekSegment = findSegmentWithEntry(entryId, errorRule);
                 moveToValidSegment();
```
[the remark this answers is not in the document]
Comment (manjujames) [reply src/frontend/org/voltdb/utils/PersistentBinaryDeque.java]: May be, I will look into it.

File: src/frontend/org/voltdb/utils/PersistentBinaryDeque.java (hunk 1 above)
Comment (manjujames) [reply src/frontend/org/voltdb/utils/PersistentBinaryDeque.java]: Indexes are monotonically increasing. That is the assumption in the code.

File: src/frontend/org/voltdb/utils/PersistentBinaryDeque.java (hunk 2 above)
Comment (manjujames) [reply src/frontend/org/voltdb/utils/PersistentBinaryDeque.java]: Here invalid key is an index that doesn't exist in the PBD. It is used to stop the search while going through indices. INVALID_KEY will not work in this case. I could change this var to invalid index to clarify it.

File: src/frontend/org/voltdb/utils/PersistentBinaryDeque.java (hunk 3 above)
Comment (manjujames) [reply src/frontend/org/voltdb/utils/PersistentBinaryDeque.java]: Yes, thanks!

File: tests/frontend/org/voltdb/export/TestExportDanglingSources.java [hunk 7]
```diff
@@ -1,194 +0,0 @@
...
-import org.voltdb.export.TestExportBaseSocketExport.ServerListener;
-import org.voltdb.regressionsuites.LocalCluster;
-
-import com.google_voltpatches.common.collect.ImmutableList;
-import com.google_voltpatches.common.collect.Lists;
-
-/**
- * Test handling of dangling data sources
- */
-public class TestExportDanglingSources extends ExportLocalClusterBase {
```
Comment (n3world) [inline tests/frontend/org/voltdb/export/TestExportDanglingSources.java]: Why is this test being removed? Isn't it still testing a valid case when a system has a dangle pbd that the pbd can be drained and deleted?

File: src/frontend/org/voltdb/utils/PBDRegularSegment.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (manjujames) [reply src/frontend/org/voltdb/utils/PBDRegularSegment.java]: Actually, I had the Long in an intermediate stage. Now I think I can get rid of it and still keep the API clean.

File: src/frontend/org/voltdb/utils/PersistentBinaryDeque.java [hunk 8]
```diff
@@ -783,8 +756,7 @@ private long getPreviousSegmentId(File file) {
      */
     private void parseFiles(boolean deleteExisting) throws IOException {
 
-        HashMap<Long, PbdSegmentName> filesById = new HashMap<>();
-        PairSequencer<Long> sequencer = new PairSequencer<>();
```
[the remark this answers is not in the document]
Comment (manjujames) [reply src/frontend/org/voltdb/utils/PersistentBinaryDeque.java]: Thanks, I will remove it.

File: tests/frontend/org/voltdb/export/TestExportDanglingSources.java (hunk 7 above)
Comment (manjujames) [reply tests/frontend/org/voltdb/export/TestExportDanglingSources.java]: In my testing, this was not creating dangling sources and the test was always failing. So Richard recommended that we remove this test.

Comment (n3world) [reply tests/frontend/org/voltdb/export/TestExportDanglingSources.java]: Interesting because it was creating dangling source and found a bug in my recent change when I removed the ad files
